controller/grpc.go

```diff
@@ -0,0 +1,1014 @@
+		return nil, err
+	}
+
+	return handler(ctx, req)
+}
+
+func (g *grpcAPI) Status(context.Context, *empty.Empty) (*protobuf.StatusResponse, error) {
+	healthy := true
+	if err := g.db.Exec("ping"); err != nil {
+		healthy = false
+	}
+	return protobuf.NewStatusResponse(healthy, nil), nil
+}
+
+func (g *grpcAPI) listApps(req *protobuf.StreamAppsRequest) ([]*protobuf.App, *data.PageToken, error) {
+	pageSize := int(req.GetPageSize())
+	pageToken, err := data.ParsePageToken(req.PageToken)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if pageSize > 0 {
+		pageToken.Size = pageSize
+	} else {
+		pageSize = pageToken.Size
+	}
+
+	appIDs := protobuf.ParseIDsFromNameFilters(req.GetNameFilters(), "apps")
+	ctApps, nextPageToken, err := g.appRepo.ListPage(data.ListAppOptions{
+		PageToken:    *pageToken,
+		AppIDs:       appIDs,
+		LabelFilters: protobuf.NewControllerLabelFilters(req.GetLabelFilters()),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if pageSize == 0 {
+		pageSize = len(ctApps)
+	}
+
+	if len(appIDs) == 0 {
+		appIDs = nil
+	}
+
+	apps := make([]*protobuf.App, 0, pageSize)
+	n := 0
+	for _, a := range ctApps {
+		apps = append(apps, protobuf.NewApp(a))
+		n++
+
+		if n == pageSize {
+			break
+		}
+	}
+
+	return apps, nextPageToken, nil
+}
+
+func (g *grpcAPI) StreamApps(req *protobuf.StreamAppsRequest, stream protobuf.Controller_StreamAppsServer) error {
+	unary := !(req.StreamUpdates || req.StreamCreates)
+
+	var apps []*protobuf.App
+	var nextPageToken *data.PageToken
+	var appsMtx sync.RWMutex
+	refreshApps := func() error {
+		appsMtx.Lock()
+		defer appsMtx.Unlock()
+		var err error
+		apps, nextPageToken, err = g.listApps(req)
+		return err
+	}
+
+	sendResponse := func() {
+		appsMtx.RLock()
+		stream.Send(&protobuf.StreamAppsResponse{
+			Apps:          apps,
+			NextPageToken: nextPageToken.String(),
+			PageComplete:  true,
+		})
+		appsMtx.RUnlock()
+	}
+
+	var sub *grpcEventListener
+	var err error
+	if !unary {
+		appIDs := protobuf.ParseIDsFromNameFilters(req.GetNameFilters(), "apps")
+		sub, err = g.subscribeEvents(appIDs, []ct.EventType{ct.EventTypeApp, ct.EventTypeAppDeletion, ct.EventTypeAppRelease}, nil)
+		if err != nil {
+			return protobuf.NewError(err, err.Error())
+		}
+		defer sub.Close()
+	}
+
+	if err := refreshApps(); err != nil {
+		return protobuf.NewError(err, err.Error())
+	}
+	sendResponse()
+	if unary {
+		return nil
+	}
+
+	maybeSendApp := func(event *ct.Event, app *protobuf.App) {
+		shouldSend := false
+		if (req.StreamCreates && event.Op == ct.EventOpCreate) || (req.StreamUpdates && event.Op == ct.EventOpUpdate) || (req.StreamUpdates && event.ObjectType == ct.EventTypeAppRelease) {
+			shouldSend = protobuf.MatchLabelFilters(app.Labels, req.GetLabelFilters())
+		}
+		if shouldSend {
+			stream.Send(&protobuf.StreamAppsResponse{
+				Apps: []*protobuf.App{app},
+			})
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			event, ok := <-sub.Events
```

Comment:
Should there be a select on the incoming stream's `ctx.Done()`  to detect disconnected clients?